KubernetesClientException when trying to open a namespace on a cluster that is NOT running (some older cluster kept in kubeconfig probably)

related to #124, #103

Steps:
1. ASSERT: have cluster/context in kubeconf that doesn't exist any more
1. EXEC: pick "Set as Current Cluster" in ctx menu for this cluster/context
1. EXEC: unfold cluster/context & "Namespaces" (category)

Result:
"Namespaces" shows a child node with an error (expected), but then a notification pops up in the bottom right corner of the IDE which reports an exception:
![image.png](https://images.zenhubusercontent.com/5c924840b65bca7678bd233e/211b2c75-2776-4ec9-9fdf-122f90f27769)
```
io.fabric8.kubernetes.client.KubernetesClientException: Operation: [list]  for kind: [Namespace]  with name: [null]  in namespace: [default]  failed.
	at io.fabric8.kubernetes.client.KubernetesClientException.launderThrowable(KubernetesClientException.java:64)
	at io.fabric8.kubernetes.client.KubernetesClientException.launderThrowable(KubernetesClientException.java:72)
	at io.fabric8.kubernetes.client.dsl.base.BaseOperation.listRequestHelper(BaseOperation.java:170)
	at io.fabric8.kubernetes.client.dsl.base.BaseOperation.list(BaseOperation.java:640)
	at io.fabric8.kubernetes.client.dsl.base.BaseOperation.list(BaseOperation.java:79)
	at org.jboss.tools.intellij.kubernetes.model.resource.NonNamespacedResourcesProvider.loadAllResources(NonNamespacedResourcesProvider.kt:36)
	at org.jboss.tools.intellij.kubernetes.model.resource.NonNamespacedResourcesProvider.getAllResources(NonNamespacedResourcesProvider.kt:28)
	at org.jboss.tools.intellij.kubernetes.model.context.ActiveContext.getAllResources(ActiveContext.kt:132)
	at org.jboss.tools.intellij.kubernetes.model.context.ActiveContext.access$getAllResources(ActiveContext.kt:58)
	at org.jboss.tools.intellij.kubernetes.model.context.ActiveContext$namespacedProviders$2.invoke(ActiveContext.kt:72)
	at org.jboss.tools.intellij.kubernetes.model.context.ActiveContext$namespacedProviders$2.invoke(ActiveContext.kt:58)
	at kotlin.SynchronizedLazyImpl.getValue(LazyJVM.kt:74)
	at org.jboss.tools.intellij.kubernetes.model.context.ActiveContext.getNamespacedProviders(ActiveContext.kt)
	at org.jboss.tools.intellij.kubernetes.model.context.ActiveContext.watch(ActiveContext.kt:222)
	at org.jboss.tools.intellij.kubernetes.model.ResourceModel.watch(ResourceModel.kt:113)
	at org.jboss.tools.intellij.kubernetes.tree.TreeStructure$FolderDescriptor.watchResources(TreeStructure.kt:199)
	at org.jboss.tools.intellij.kubernetes.tree.ResourceWatchController$install$1.treeExpanded(ResourceWatchController.kt:28)
	at javax.swing.JTree.fireTreeExpanded(JTree.java:2775)
	at javax.swing.JTree.setExpandedState(JTree.java:3629)
	at javax.swing.JTree.expandPath(JTree.java:2212)
	at javax.swing.plaf.basic.BasicTreeUI.toggleExpandState(BasicTreeUI.java:2286)
	at javax.swing.plaf.basic.BasicTreeUI.selectPathForEvent(BasicTreeUI.java:2395)
	at javax.swing.plaf.basic.BasicTreeUI$Handler.handleSelection(BasicTreeUI.java:3609)
	at javax.swing.plaf.basic.BasicTreeUI$Handler.mousePressed(BasicTreeUI.java:3548)
	at com.intellij.util.ui.MouseEventAdapter.mousePressed(MouseEventAdapter.java:33)
	at java.awt.AWTEventMulticaster.mousePressed(AWTEventMulticaster.java:279)
	at java.awt.AWTEventMulticaster.mousePressed(AWTEventMulticaster.java:279)
	at java.awt.AWTEventMulticaster.mousePressed(AWTEventMulticaster.java:279)
	at java.awt.AWTEventMulticaster.mousePressed(AWTEventMulticaster.java:279)
	at java.awt.Component.processMouseEvent(Component.java:6545)
	at javax.swing.JComponent.processMouseEvent(JComponent.java:3325)
	at com.intellij.ui.treeStructure.Tree.processMouseEvent(Tree.java:409)
	at java.awt.Component.processEvent(Component.java:6313)
	at java.awt.Container.processEvent(Container.java:2237)
	at java.awt.Component.dispatchEventImpl(Component.java:4903)
	at java.awt.Container.dispatchEventImpl(Container.java:2295)
	at java.awt.Component.dispatchEvent(Component.java:4725)
	at java.awt.LightweightDispatcher.retargetMouseEvent(Container.java:4889)
	at java.awt.LightweightDispatcher.processMouseEvent(Container.java:4523)
	at java.awt.LightweightDispatcher.dispatchEvent(Container.java:4467)
	at java.awt.Container.dispatchEventImpl(Container.java:2281)
	at java.awt.Window.dispatchEventImpl(Window.java:2746)
	at java.awt.Component.dispatchEvent(Component.java:4725)
	at java.awt.EventQueue.dispatchEventImpl(EventQueue.java:764)
	at java.awt.EventQueue.access$500(EventQueue.java:98)
	at java.awt.EventQueue$3.run(EventQueue.java:715)
	at java.awt.EventQueue$3.run(EventQueue.java:709)
	at java.security.AccessController.doPrivileged(Native Method)
	at java.security.ProtectionDomain$JavaSecurityAccessImpl.doIntersectionPrivilege(ProtectionDomain.java:80)
	at java.security.ProtectionDomain$JavaSecurityAccessImpl.doIntersectionPrivilege(ProtectionDomain.java:90)
	at java.awt.EventQueue$4.run(EventQueue.java:737)
	at java.awt.EventQueue$4.run(EventQueue.java:735)
	at java.security.AccessController.doPrivileged(Native Method)
	at java.security.ProtectionDomain$JavaSecurityAccessImpl.doIntersectionPrivilege(ProtectionDomain.java:80)
	at java.awt.EventQueue.dispatchEvent(EventQueue.java:734)
	at com.intellij.ide.IdeEventQueue.defaultDispatchEvent(IdeEventQueue.java:719)
	at com.intellij.ide.IdeEventQueue._dispatchEvent(IdeEventQueue.java:664)
	at com.intellij.ide.IdeEventQueue.dispatchEvent(IdeEventQueue.java:363)
	at java.awt.EventDispatchThread.pumpOneEventForFilters(EventDispatchThread.java:201)
	at java.awt.EventDispatchThread.pumpEventsForFilter(EventDispatchThread.java:116)
	at java.awt.EventDispatchThread.pumpEventsForHierarchy(EventDispatchThread.java:105)
	at java.awt.EventDispatchThread.pumpEvents(EventDispatchThread.java:101)
	at java.awt.EventDispatchThread.pumpEvents(EventDispatchThread.java:93)
	at java.awt.EventDispatchThread.run(EventDispatchThread.java:82)
Caused by: java.net.NoRouteToHostException: No route to host (Host unreachable)
	at java.net.PlainSocketImpl.socketConnect(Native Method)
	at java.net.AbstractPlainSocketImpl.doConnect(AbstractPlainSocketImpl.java:350)
	at java.net.AbstractPlainSocketImpl.connectToAddress(AbstractPlainSocketImpl.java:206)
	at java.net.AbstractPlainSocketImpl.connect(AbstractPlainSocketImpl.java:188)
	at java.net.SocksSocketImpl.connect(SocksSocketImpl.java:392)
	at java.net.Socket.connect(Socket.java:589)
	at okhttp3.internal.platform.Platform.connectSocket(Platform.java:129)
	at okhttp3.internal.connection.RealConnection.connectSocket(RealConnection.java:247)
	at okhttp3.internal.connection.RealConnection.connect(RealConnection.java:167)
	at okhttp3.internal.connection.StreamAllocation.findConnection(StreamAllocation.java:258)
	at okhttp3.internal.connection.StreamAllocation.findHealthyConnection(StreamAllocation.java:135)
	at okhttp3.internal.connection.StreamAllocation.newStream(StreamAllocation.java:114)
	at okhttp3.internal.connection.ConnectInterceptor.intercept(ConnectInterceptor.java:42)
	at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:147)
	at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:121)
	at okhttp3.internal.cache.CacheInterceptor.intercept(CacheInterceptor.java:93)
	at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:147)
	at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:121)
	at okhttp3.internal.http.BridgeInterceptor.intercept(BridgeInterceptor.java:93)
	at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:147)
	at okhttp3.internal.http.RetryAndFollowUpInterceptor.intercept(RetryAndFollowUpInterceptor.java:127)
	at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:147)
	at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:121)
	at io.fabric8.kubernetes.client.utils.BackwardsCompatibilityInterceptor.intercept(BackwardsCompatibilityInterceptor.java:134)
	at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:147)
	at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:121)
	at io.fabric8.kubernetes.client.utils.ImpersonatorInterceptor.intercept(ImpersonatorInterceptor.java:68)
	at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:147)
	at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:121)
	at io.fabric8.kubernetes.client.utils.HttpClientUtils.lambda$createHttpClient$3(HttpClientUtils.java:147)
	at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:147)
	at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:121)
	at okhttp3.RealCall.getResponseWithInterceptorChain(RealCall.java:257)
	at okhttp3.RealCall.execute(RealCall.java:93)
	at io.fabric8.kubernetes.client.dsl.base.OperationSupport.handleResponse(OperationSupport.java:490)
	at io.fabric8.kubernetes.client.dsl.base.OperationSupport.handleResponse(OperationSupport.java:451)
	at io.fabric8.kubernetes.client.dsl.base.OperationSupport.handleResponse(OperationSupport.java:433)
	at io.fabric8.kubernetes.client.dsl.base.BaseOperation.listRequestHelper(BaseOperation.java:166)
	... 61 more
```

This happens apparently when the server isn't up, right?

> Caused by: java.net.NoRouteToHostException: No route to host (Host unreachable)

ps. I'd be grateful for steps and more details so that I can reproduce the different bugs that you found

--- a/src/main/kotlin/org/jboss/tools/intellij/kubernetes/actions/Extensions.kt
+++ b/src/main/kotlin/org/jboss/tools/intellij/kubernetes/actions/Extensions.kt
@@ -12,6 +12,7 @@ package org.jboss.tools.intellij.kubernetes.actions
 
 import com.intellij.openapi.actionSystem.AnAction
 import com.intellij.openapi.components.ServiceManager
+import io.fabric8.kubernetes.api.model.HasMetadata
 import org.jboss.tools.intellij.kubernetes.model.IResourceModel
 import org.jboss.tools.intellij.kubernetes.tree.TreeStructure
 import javax.swing.tree.DefaultMutableTreeNode
@@ -28,3 +29,9 @@ fun Any?.getDescriptor(): TreeStructure.Descriptor<*>? {
 fun <T> Any.getElement(): T? {
     return (this as? DefaultMutableTreeNode)?.getDescriptor()?.element as? T
 }
+
+fun TreeStructure.Descriptor<*>.getKind(): String? {
+    val element = this?.element
+    return (element as? TreeStructure.Folder)?.kind?.kind ?: (element as? HasMetadata)?.kind
+}
+

--- a/src/main/kotlin/org/jboss/tools/intellij/kubernetes/actions/Extensions.kt
+++ b/src/main/kotlin/org/jboss/tools/intellij/kubernetes/actions/Extensions.kt
@@ -12,9 +12,17 @@ package org.jboss.tools.intellij.kubernetes.actions
 
 import com.intellij.openapi.actionSystem.AnAction
 import com.intellij.openapi.components.ServiceManager
+import com.intellij.openapi.diagnostic.logger
+import com.intellij.openapi.progress.ProgressIndicator
+import com.intellij.openapi.progress.ProgressManager
+import com.intellij.openapi.progress.Progressive
+import com.intellij.openapi.progress.Task
+import com.redhat.devtools.intellij.common.actions.StructureTreeAction
 import io.fabric8.kubernetes.api.model.HasMetadata
 import org.jboss.tools.intellij.kubernetes.model.IResourceModel
+import org.jboss.tools.intellij.kubernetes.tree.ResourceWatchController
 import org.jboss.tools.intellij.kubernetes.tree.TreeStructure
+import org.jetbrains.annotations.NotNull
 import javax.swing.tree.DefaultMutableTreeNode
 
 
@@ -35,3 +43,13 @@ fun TreeStructure.Descriptor<*>.getKind(): String? {
     return (element as? TreeStructure.Folder)?.kind?.kind ?: (element as? HasMetadata)?.kind
 }
 
+fun AnAction.run(title: String, canBeCancelled: Boolean, runnable: Progressive) {
+    ProgressManager.getInstance().run(object :
+        Task.Backgroundable(null, title, canBeCancelled){
+
+        override fun run(@NotNull progress: ProgressIndicator) {
+            runnable.run(progress)
+        }
+    })
+
+}

--- a/src/main/kotlin/org/jboss/tools/intellij/kubernetes/actions/RefreshAction.kt
+++ b/src/main/kotlin/org/jboss/tools/intellij/kubernetes/actions/RefreshAction.kt
@@ -11,12 +11,19 @@
 package org.jboss.tools.intellij.kubernetes.actions
 
 import com.intellij.openapi.actionSystem.AnActionEvent
+import com.intellij.openapi.diagnostic.logger
 import com.redhat.devtools.intellij.common.actions.StructureTreeAction
+import org.jboss.tools.intellij.kubernetes.tree.ResourceWatchController
 import javax.swing.tree.TreePath
 
 class RefreshAction: StructureTreeAction(Any::class.java) {
 
     override fun actionPerformed(event: AnActionEvent?, path: TreePath?, selectedNode: Any?) {
-        selectedNode?.getDescriptor()?.invalidate()
+        val descriptor = selectedNode?.getDescriptor() ?: return
+        try {
+            descriptor.invalidate()
+        } catch(e: Exception) {
+            logger<ResourceWatchController>().warn("Could not refresh $descriptor resources.", e)
+        }
     }
 }

--- a/src/main/kotlin/org/jboss/tools/intellij/kubernetes/actions/RefreshAction.kt
+++ b/src/main/kotlin/org/jboss/tools/intellij/kubernetes/actions/RefreshAction.kt
@@ -12,18 +12,22 @@ package org.jboss.tools.intellij.kubernetes.actions
 
 import com.intellij.openapi.actionSystem.AnActionEvent
 import com.intellij.openapi.diagnostic.logger
+import com.intellij.openapi.progress.Progressive
 import com.redhat.devtools.intellij.common.actions.StructureTreeAction
 import org.jboss.tools.intellij.kubernetes.tree.ResourceWatchController
 import javax.swing.tree.TreePath
 
-class RefreshAction: StructureTreeAction(Any::class.java) {
+class RefreshAction : StructureTreeAction(Any::class.java) {
 
-    override fun actionPerformed(event: AnActionEvent?, path: TreePath?, selectedNode: Any?) {
-        val descriptor = selectedNode?.getDescriptor() ?: return
-        try {
-            descriptor.invalidate()
-        } catch(e: Exception) {
-            logger<ResourceWatchController>().warn("Could not refresh $descriptor resources.", e)
-        }
-    }
+	override fun actionPerformed(event: AnActionEvent?, path: TreePath?, selectedNode: Any?) {
+		val descriptor = selectedNode?.getDescriptor() ?: return
+		run("Refreshing $selectedNode...", true,
+			Progressive {
+				try {
+					descriptor.invalidate()
+				} catch (e: Exception) {
+					logger<ResourceWatchController>().warn("Could not refresh $descriptor resources.", e)
+				}
+			})
+	}
 }

--- a/src/main/kotlin/org/jboss/tools/intellij/kubernetes/actions/SetAsCurrentClusterAction.kt
+++ b/src/main/kotlin/org/jboss/tools/intellij/kubernetes/actions/SetAsCurrentClusterAction.kt
@@ -12,6 +12,7 @@ package org.jboss.tools.intellij.kubernetes.actions
 
 import com.intellij.openapi.actionSystem.AnActionEvent
 import com.intellij.openapi.diagnostic.logger
+import com.intellij.openapi.progress.Progressive
 import com.redhat.devtools.intellij.common.actions.StructureTreeAction
 import org.jboss.tools.intellij.kubernetes.model.context.IContext
 import org.jboss.tools.intellij.kubernetes.tree.ResourceWatchController
@@ -21,11 +22,15 @@ class SetAsCurrentClusterAction: StructureTreeAction(IContext::class.java) {
 
     override fun actionPerformed(event: AnActionEvent?, path: TreePath?, selectedNode: Any?) {
         val context: IContext = selectedNode?.getElement() ?: return
-        try {
-            getResourceModel()?.setCurrentContext(context)
-        } catch(e: Exception) {
-            logger<ResourceWatchController>().warn(
-                "Could not set current context to ${context.context.name}.", e)
-        }
+        run("Setting $context as current cluster...", true,
+            Progressive {
+                try {
+                    getResourceModel()?.setCurrentContext(context)
+                } catch (e: Exception) {
+                    logger<ResourceWatchController>().warn(
+                        "Could not set current context to ${context.context.name}.", e
+                    )
+                }
+            })
     }
 }

--- a/src/main/kotlin/org/jboss/tools/intellij/kubernetes/actions/SetAsCurrentClusterAction.kt
+++ b/src/main/kotlin/org/jboss/tools/intellij/kubernetes/actions/SetAsCurrentClusterAction.kt
@@ -11,14 +11,21 @@
 package org.jboss.tools.intellij.kubernetes.actions
 
 import com.intellij.openapi.actionSystem.AnActionEvent
+import com.intellij.openapi.diagnostic.logger
 import com.redhat.devtools.intellij.common.actions.StructureTreeAction
 import org.jboss.tools.intellij.kubernetes.model.context.IContext
+import org.jboss.tools.intellij.kubernetes.tree.ResourceWatchController
 import javax.swing.tree.TreePath
 
 class SetAsCurrentClusterAction: StructureTreeAction(IContext::class.java) {
 
     override fun actionPerformed(event: AnActionEvent?, path: TreePath?, selectedNode: Any?) {
         val context: IContext = selectedNode?.getElement() ?: return
-        getResourceModel()?.setCurrentContext(context)
+        try {
+            getResourceModel()?.setCurrentContext(context)
+        } catch(e: Exception) {
+            logger<ResourceWatchController>().warn(
+                "Could not set current context to ${context.context.name}.", e)
+        }
     }
 }

--- a/src/main/kotlin/org/jboss/tools/intellij/kubernetes/actions/UseNamespaceAction.kt
+++ b/src/main/kotlin/org/jboss/tools/intellij/kubernetes/actions/UseNamespaceAction.kt
@@ -11,11 +11,13 @@
 package org.jboss.tools.intellij.kubernetes.actions
 
 import com.intellij.openapi.actionSystem.AnActionEvent
+import com.intellij.openapi.diagnostic.logger
 import com.intellij.openapi.progress.ProgressIndicator
 import com.intellij.openapi.progress.ProgressManager
 import com.intellij.openapi.progress.Task
 import com.redhat.devtools.intellij.common.actions.StructureTreeAction
 import io.fabric8.kubernetes.api.model.Namespace
+import org.jboss.tools.intellij.kubernetes.tree.ResourceWatchController
 import org.jetbrains.annotations.NotNull
 import javax.swing.tree.TreePath
 
@@ -24,11 +26,16 @@ class UseNamespaceAction: StructureTreeAction(Namespace::class.java) {
     override fun actionPerformed(event: AnActionEvent?, path: TreePath?, selectedNode: Any?) {
         val namespace: Namespace = selectedNode?.getElement() ?: return
         val model = getResourceModel() ?: return
-        val manager = ProgressManager.getInstance()
-        manager.run(object : Task.Backgroundable(null, "Switching Namespace...", true) {
+        ProgressManager.getInstance().run(object :
+            Task.Backgroundable(null, "Switching namespace...", true) {
 
             override fun run(@NotNull progress: ProgressIndicator) {
+                try {
                     model.setCurrentNamespace(namespace.metadata.name)
+                } catch(e: Exception) {
+                    logger<ResourceWatchController>().warn(
+                        "Could not use namespace ${namespace.metadata.name}.", e)
+                }
             }
         })
     }

--- a/src/main/kotlin/org/jboss/tools/intellij/kubernetes/actions/UseNamespaceAction.kt
+++ b/src/main/kotlin/org/jboss/tools/intellij/kubernetes/actions/UseNamespaceAction.kt
@@ -14,6 +14,7 @@ import com.intellij.openapi.actionSystem.AnActionEvent
 import com.intellij.openapi.diagnostic.logger
 import com.intellij.openapi.progress.ProgressIndicator
 import com.intellij.openapi.progress.ProgressManager
+import com.intellij.openapi.progress.Progressive
 import com.intellij.openapi.progress.Task
 import com.redhat.devtools.intellij.common.actions.StructureTreeAction
 import io.fabric8.kubernetes.api.model.Namespace
@@ -26,17 +27,15 @@ class UseNamespaceAction: StructureTreeAction(Namespace::class.java) {
     override fun actionPerformed(event: AnActionEvent?, path: TreePath?, selectedNode: Any?) {
         val namespace: Namespace = selectedNode?.getElement() ?: return
         val model = getResourceModel() ?: return
-        ProgressManager.getInstance().run(object :
-            Task.Backgroundable(null, "Switching namespace...", true) {
-
-            override fun run(@NotNull progress: ProgressIndicator) {
+        run("Switching namespace $namespace...", true,
+            Progressive {
                 try {
                     model.setCurrentNamespace(namespace.metadata.name)
-                } catch(e: Exception) {
+                } catch (e: Exception) {
                     logger<ResourceWatchController>().warn(
-                        "Could not use namespace ${namespace.metadata.name}.", e)
+                        "Could not use namespace ${namespace.metadata.name}.", e
+                    )
                 }
-            }
-        })
+            })
     }
 }

--- a/src/main/kotlin/org/jboss/tools/intellij/kubernetes/actions/UseProjectAction.kt
+++ b/src/main/kotlin/org/jboss/tools/intellij/kubernetes/actions/UseProjectAction.kt
@@ -11,14 +11,30 @@
 package org.jboss.tools.intellij.kubernetes.actions
 
 import com.intellij.openapi.actionSystem.AnActionEvent
+import com.intellij.openapi.diagnostic.logger
+import com.intellij.openapi.progress.ProgressIndicator
+import com.intellij.openapi.progress.ProgressManager
+import com.intellij.openapi.progress.Task
 import com.redhat.devtools.intellij.common.actions.StructureTreeAction
 import io.fabric8.openshift.api.model.Project
+import org.jboss.tools.intellij.kubernetes.tree.ResourceWatchController
+import org.jetbrains.annotations.NotNull
 import javax.swing.tree.TreePath
 
-class UseProjectAction: StructureTreeAction(Project::class.java) {
+class UseProjectAction : StructureTreeAction(Project::class.java) {
 
-    override fun actionPerformed(event: AnActionEvent?, path: TreePath?, selectedNode: Any?) {
-        val project: Project = selectedNode?.getElement() ?: return
-        getResourceModel()?.setCurrentNamespace(project.metadata.name)
-    }
+	override fun actionPerformed(event: AnActionEvent?, path: TreePath?, selectedNode: Any?) {
+		val project: Project = selectedNode?.getElement() ?: return
+		ProgressManager.getInstance().run(object :
+			Task.Backgroundable(null, "Switching project...", true) {
+			override fun run(@NotNull progress: ProgressIndicator) {
+				try {
+					getResourceModel()?.setCurrentNamespace(project.metadata.name)
+				} catch (e: Exception) {
+					logger<ResourceWatchController>().warn(
+						"Could not use namespace ${project.metadata.name}.", e)
+				}
+			}
+		})
+	}
 }

--- a/src/main/kotlin/org/jboss/tools/intellij/kubernetes/actions/UseProjectAction.kt
+++ b/src/main/kotlin/org/jboss/tools/intellij/kubernetes/actions/UseProjectAction.kt
@@ -14,6 +14,7 @@ import com.intellij.openapi.actionSystem.AnActionEvent
 import com.intellij.openapi.diagnostic.logger
 import com.intellij.openapi.progress.ProgressIndicator
 import com.intellij.openapi.progress.ProgressManager
+import com.intellij.openapi.progress.Progressive
 import com.intellij.openapi.progress.Task
 import com.redhat.devtools.intellij.common.actions.StructureTreeAction
 import io.fabric8.openshift.api.model.Project
@@ -25,16 +26,16 @@ class UseProjectAction : StructureTreeAction(Project::class.java) {
 
 	override fun actionPerformed(event: AnActionEvent?, path: TreePath?, selectedNode: Any?) {
 		val project: Project = selectedNode?.getElement() ?: return
-		ProgressManager.getInstance().run(object :
-			Task.Backgroundable(null, "Switching project...", true) {
-			override fun run(@NotNull progress: ProgressIndicator) {
+		val projectName = project.metadata.name
+		run("Using project $projectName...", true,
+			Progressive {
 				try {
-					getResourceModel()?.setCurrentNamespace(project.metadata.name)
+					getResourceModel()?.setCurrentNamespace(projectName)
 				} catch (e: Exception) {
 					logger<ResourceWatchController>().warn(
-						"Could not use namespace ${project.metadata.name}.", e)
+						"Could not use namespace $projectName.", e
+					)
 				}
-			}
-		})
+			})
 	}
 }

--- a/src/main/kotlin/org/jboss/tools/intellij/kubernetes/model/ResourceWatch.kt
+++ b/src/main/kotlin/org/jboss/tools/intellij/kubernetes/model/ResourceWatch.kt
@@ -128,7 +128,7 @@ open class ResourceWatch(
         override fun run() {
             try {
                 logger<ResourceWatcher>().debug("Watching $kind resources.")
-                val watch: Watch? = createWatch()
+                val watch: Watch? = watch()
                 saveWatch(watch)
             } catch (e: Exception) {
                 watches.remove(kind) // remove placeholder
@@ -136,7 +136,7 @@ open class ResourceWatch(
             }
         }
 
-        private fun createWatch(): Watch? {
+        private fun watch(): Watch? {
             return if (watchable == null) {
               return null
             } else {

--- a/src/main/kotlin/org/jboss/tools/intellij/kubernetes/model/util/Logger.kt
+++ b/src/main/kotlin/org/jboss/tools/intellij/kubernetes/model/util/Logger.kt
@@ -23,3 +23,6 @@ fun Logger.debugWithClass(clazz: Class<*>, message: String) {
 fun Logger.debugWithClass(clazz: Class<*>, message: String, t: Throwable?) {
 	this.debug("${clazz.simpleName}: $message", t)
 }
+
+
+

--- a/src/main/kotlin/org/jboss/tools/intellij/kubernetes/tree/ResourceWatchController.kt
+++ b/src/main/kotlin/org/jboss/tools/intellij/kubernetes/tree/ResourceWatchController.kt
@@ -10,6 +10,7 @@
  ******************************************************************************/
 package org.jboss.tools.intellij.kubernetes.tree
 
+import com.intellij.openapi.diagnostic.logger
 import org.jboss.tools.intellij.kubernetes.actions.getDescriptor
 import javax.swing.JTree
 import javax.swing.event.TreeExpansionEvent
@@ -25,7 +26,11 @@ object ResourceWatchController {
 		tree.addTreeExpansionListener(object: TreeExpansionListener {
 			override fun treeExpanded(event: TreeExpansionEvent) {
 				val descriptor = (event.path.lastPathComponent as DefaultMutableTreeNode).getDescriptor()
-				descriptor?.watchResources()
+				try {
+					descriptor?.watchResources()
+				} catch(e: Exception) {
+					logger<ResourceWatchController>().warn("Could not watch ${descriptor?.element} resources.", e)
+				}
 			}
 
 			override fun treeCollapsed(event: TreeExpansionEvent?) {

--- a/src/main/resources/META-INF/plugin.xml
+++ b/src/main/resources/META-INF/plugin.xml
@@ -10,19 +10,15 @@
                     anchor="left"
                     factoryClass="org.jboss.tools.intellij.kubernetes.KubernetesToolWindowFactory"
                     icon="/META-INF/pluginIcon13x13.svg"/>
-    </extensions>
-
-    <extensions defaultExtensionNs="com.intellij">
         <applicationService serviceInterface="org.jboss.tools.intellij.kubernetes.model.IResourceModel"
                         serviceImplementation="org.jboss.tools.intellij.kubernetes.model.ResourceModel" />
     </extensions>
 
     <extensionPoints>
-        <extensionPoint name="resourceProvider" interface="org.jboss.tools.intellij.kubernetes.model.resource.IResourcesProviderFactory"/>
-    </extensionPoints>
-
-    <extensionPoints>
-        <extensionPoint name="structureContribution" interface="org.jboss.tools.intellij.kubernetes.tree.ITreeStructureContributionFactory"/>
+        <extensionPoint name="resourceProvider"
+                        interface="org.jboss.tools.intellij.kubernetes.model.resource.IResourcesProviderFactory"/>
+        <extensionPoint name="structureContribution"
+                        interface="org.jboss.tools.intellij.kubernetes.tree.ITreeStructureContributionFactory"/>
     </extensionPoints>
 
     <actions>

